graph/src/components/ethereum/types.rs

```diff
@@ -67,6 +67,42 @@ pub struct EthereumBlockWithCalls {
     pub calls: Option<Vec<EthereumCall>>,
 }
 
+impl EthereumBlockWithCalls {
+    /// Given an `EthereumCall`, check within receipts if that transaction was successful.
+    pub fn transaction_for_call_succeeded(&self, call: &EthereumCall) -> anyhow::Result<bool> {
+        let call_transaction_hash = call.transaction_hash.ok_or(anyhow::anyhow!(
+            "failed to find a transaction for this call"
+        ))?;
+
+        let receipt = self
+            .ethereum_block
+            .transaction_receipts
+            .iter()
+            .find(|txn| txn.transaction_hash == call_transaction_hash)
+            .ok_or(anyhow::anyhow!(
+                "failed to find the receipt for this transaction"
+            ))?;
+
+        let transaction = self
+            .ethereum_block
+            .block
+            .transaction_for_call(&call)
+            .ok_or(anyhow::anyhow!(
+                "failed to find the transaction for this call"
+            ))?;
+
+        // assume the transaction failed if all gas was used
+        if receipt
+            .gas_used
+            .ok_or(anyhow::anyhow!("Running in light client mode)"))?
+            >= transaction.gas
+        {
+            return Ok(false);
+        }
+        Ok(matches!(receipt.status, Some(x) if x == web3::types::U64::from(1)))
```

Comment:
This logic seems duplicated with `fn is_transaction_successful`, is there a way to de-duplicate it?

Reply:
Placed the common logic in the new `evaluate_transaction_status` function.